cpp/cograph_recognition/cograph/Recognition.cpp

```diff
@@ -0,0 +1,233 @@
+                u.prev = nullptr;
+                u.next = w.head_of_list_of_children;
+                w.head_of_list_of_children.prev = u;
+                w.head_of_list_of_children = u;
+            }
+            //else u is head
+         }
+    }
+}
+void Mark( CoNode &x){
+    queue<CoNode> marked_with_d_equal_to_md;
+    mark_count = 0;
+    mark_and_unmarked_count = 0;
+    mark_ever_count = 0;
+    for(auto u : T.outEdges(x)){
+        u.mark();
+        mark_ever_count++;
+        mark_count++;
+        marked_with_d_equal_to_md.push(u);
+    }
+    while(!marked_with_d_equal_to_md.empty()){
+        unmark( marked_with_d_equal_to_md);
+    }
+        if(mark_count){
+            if(T.getRoot().d == 1){
+                T.getRoot().mark();
+            }
+        }
+
+}
+
+void Reset_All_CoNodes(CoNode *x){
+    x -> unmark_for_new_iteration();
+    auto y = x -> head_of_list_of_children;
+    while(y != nullptr){
+        Reset_All_CoNodes(y);
+        y = y -> next;
+    }
+}
+CoNode Find_Lowest( ){
+
+}
+void Insert_x_to_CoTree(CoNode *u, CoNode *x){
+    if(u -> )
+}
+void Cograph_Recognition(NetworKit::Graph &graph){
+    G = graph;
+    vector<node>vertex;
+    vector<CoNode>covertex;
+    vector<int>pos(G.z + 1);
+    for(int i = 0; i <= G.z; i++){
+        if(G.exists(i)){
+            pos[i] = vertex.size();
+            vertex.push_back(i);
+            CoNode C = CoNode(Type::VERTEX,i);
+            covertex.push_back(C);
+        }
+    }
+
+    for(int i = 0; i <= G.z; i++){
+            if(G.exists(i)){
+                vector<CoNode> vec;
+                for(auto u : G.outEdges(i)){
+                    vec.push_back(covertex[pos[u]]);
+                }
+                covertex[pos[i]].setoutEdges(vec);
+            }
+        }
+    CoNode R(Type::ZEROONE, 1);
+    T = CoTree(&R);
+    if(vertex.size() == 0){
+        return;
+    }
+    if(vertex.size() == 1){
+        R.addchild(&covertex[0]);
+        return;
+    }
+    if(G.hasEdge(vertex[0], vertex[1])){
+        R.addchild(&covertex[0]);
+        R.addchild(&covertex[1]);
+    } else{
+        CoNode N(Type::ZEROONE, 0);
+        R.addchild(&N);
+        N.addchild(&covertex[0]);
+        N.addchild(&covertex[1]);
+    }
+    for(int i = 2; i < vertex.size()){
+        Reset_All_CoNodes(&R);
+        Mark(covertex[i]);
+        if(R.marked == Marked::MARKED_AND_UNMARKED){//all nodes of T were marked and unmarked <=> R is marked and unmarked
+            R.addchild(&covertex[i]);
+        } else if(mark_ever_count == 0){
+            if(R.d == 1){
+                (R.head_of_list_of_children) -> addchild(&covertex[i]);
+            } else{
+                CoNode R1(Type::ZEROONE, 1);
+                CoNode R2(Type::ZEROONE, 0);
+                R1.addchild(&R2);
+                R2.addchild(&R);
+                R2.addchild(&x);
+            }
+        } else{
+            CoNode u = Find_Lowest();
+            Insert_x_to_CoTree(&u, &covertex[i]);
```

Comment:
Please change to passing references and invoke `Insert_x_to_CoTree(Find_Lowest(), covertex[i]);`